src/backend/cdb/cdbllize.c

```diff
@@ -573,6 +573,18 @@ ParallelizeCorrelatedSubPlanMutator(Node *node, ParallelizeCorrelatedPlanWalkerC
 		if (ctx->movement == MOVEMENT_BROADCAST)
 		{
 			Assert (NULL != ctx->currentPlanFlow);
+
+			Plan	   *origPlan = planner_subplan_get_plan((PlannerInfo *) ctx->base.node, ctx->sp);
+			PlannerInfo *origRoot = planner_subplan_get_root((PlannerInfo *) ctx->base.node, ctx->sp);
+
+			if ((origPlan->flow->locustype == CdbLocusType_SegmentGeneral) &&
+				(contain_volatile_functions((Node *) origPlan->targetlist) ||
+				contain_volatile_functions(origRoot->parse->havingQual)))
```

Comment:
It seems that there is no reason to validate something besides target list and quals for volatile functions presence for scanPlan node, because gpdb inserts motion strictly above scan below any aggregations.

Reply:
> It seems that there is no reason to validate something besides target list and quals for volatile functions presence for scanPlan node, because gpdb inserts motion strictly above scan below any aggregations.

without it, some tests are break
```diff
--- \/home\/src\/gpdb6\/src\/test\/regress\/expected\/rpt_optimizer\.out	2022-08-30 12:58:14.829781300 +0500
+++ \/home\/src\/gpdb6\/src\/test\/regress\/results\/rpt\.out	2022-08-30 12:58:14.845781471 +0500
@@ -1557,12 +1557,12 @@
 explain (costs off, verbose) select * from t1 where a in (select random() from t where i=a group by i);
 QUERY PLAN
 ___________
- Gather Motion 3:1  (slice2; segments: 3)
+ Gather Motion 3:1  (slice1; segments: 3)
    Output: t1.a
    ->  Seq Scan on rpt.t1
          Output: t1.a
          Filter: (SubPlan 1)
-         SubPlan 1  (slice2; segments: 3)
+         SubPlan 1  (slice1; segments: 1)
            ->  GroupAggregate
                  Output: random(), t.i
                  Group Key: t.i
@@ -1571,32 +1571,28 @@
                        Filter: (t.i = t1.a)
                        ->  Materialize
                              Output: t.i, t.i
-                             ->  Broadcast Motion 1:3  (slice1; segments: 1)
+                             ->  Seq Scan on rpt.t
                                    Output: t.i, t.i
-                                   ->  Seq Scan on rpt.t
-                                         Output: t.i, t.i
-GP_IGNORE:(20 rows)
+GP_IGNORE:(18 rows)
 
 -- ensure we do not break broadcast motion
 explain (costs off, verbose) select * from t1 where a in (select random() from t where i=a);
 QUERY PLAN
 ___________
- Gather Motion 3:1  (slice2; segments: 3)
+ Gather Motion 3:1  (slice1; segments: 3)
    Output: t1.a
    ->  Seq Scan on rpt.t1
          Output: t1.a
          Filter: (SubPlan 1)
-         SubPlan 1  (slice2; segments: 3)
+         SubPlan 1  (slice1; segments: 1)
            ->  Result
                  Output: random()
                  Filter: (t.i = t1.a)
                  ->  Materialize
                        Output: t.i
-                       ->  Broadcast Motion 1:3  (slice1; segments: 1)
+                       ->  Seq Scan on rpt.t
                              Output: t.i
-                             ->  Seq Scan on rpt.t
-                                   Output: t.i
-GP_IGNORE:(17 rows)
+GP_IGNORE:(15 rows)
 
 explain (costs off, verbose) select 1 as mrs_t1 from t1 where 1 <= ALL (select i from t group by i having random() > 0);
 QUERY PLAN

======================================================================

```

Reply:
It was expected, because broadcasting of replicated tables without volatile functions is useless. Quite strange changing of segments for subplan: `SubPlan 1  (slice1; segments: 1)`

Reply:
> Quite strange changing of segments for subplan: SubPlan 1 (slice1; segments: 1)

As a note for reviewers, these changes caused by [logic](https://github.com/arenadata/gpdb/blob/2ba21dd2c9a3ddf4ae165dcc278ef5647ab68f1b/src/backend/commands/explain.c#L981-L982) of calculation subplan segments in the explain.c. FLOW_SINGLETON is always interpreted as only one segment, but scans of replicated tables has the same flow by default.